Automattic/pocket-casts-android
# add check for user preferences if user prefers unmetered connection before updating

## Description

When a user chooses the option to download "only on unmetered wifi", they essentially want to minimize mobile data usage. There's also a setting under "storage & data use" -> "Warn before using data". This warning is designed to warn the user before downloading or streaming episodes.

Currently the task that runs in the background will update podcasts even while on a metered network, which can trigger podcasts to get automatically added to the up next queue _(such as if the setting "auto download episodes added to up next" combined with the "auto add to up next" for specific podcasts)_. This will indirectly cause podcasts to be downloaded on a metered network when added to the up next queue through these settings because a podcast in the up next queue will eventually download when the podcast is played. This is mostly noticable for podcasts that automatically enqueue in the up next queue at the top instead of the bottom.

The fix for this issue hinges on keying off the option to "Warn before using data". This is co-located with the "Background Refresh" option. If the user wants to be warned about data being used, creating conditions that will create data to be used by updating in the background regardless of this setting seems to contradict the user preference. RefreshPodcastTask has been updated to use the setting for warnings (`settings.warnOnMeteredNetwork()`) to update the constraints on the work manager so that it only runs background updates while on an unmetered network. Users can disable this option if they want background updates to use data to pull updates. Since there's already a setting to control downloads "only on unmetered wifi", it has no effect for users that wish to update in the background and download on or off metered networks.

By preventing the podcast from automatically updating in the background on a metered network based on the user's preference to be warned, we help avoid setting the user up for needing a warning by not automatically getting new information about the podcasts that would require a warning in the first place.

This doesn't cover the edge case of updating while leaving an unmetered network, but I think I can live with that 😄 

Fixes #839 

## Testing Instructions

1. Load the app in the debugger
2. Set a breakpoint in the refresh podcast task [here](https://github.com/Automattic/pocket-casts-android/blob/a8bf74fd4924160c9fd6c6cf24a6608123b54c6b/modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsTask.kt#L85). Also set a breakpoint [here ](https://github.com/Automattic/pocket-casts-android/blob/a8bf74fd4924160c9fd6c6cf24a6608123b54c6b/modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/refresh/RefreshPodcastsTask.kt#L89) for the case when we disable the warning.
3. Go to "Settings - > Storage & Data Use"
4. Enable the option "Warn before using data"
5. Observe the breakpoint on line 85 being hit as well as line 89
6. Disable the option "warn before using data"
7. Observe only the breakpoint on line 89 being hit

## Screenshots or Screencast 

![228794549-3d3e542b-c28e-49a3-9387-418b839c4434](https://user-images.githubusercontent.com/12544461/228945887-ef68c3b8-0295-4923-b8d8-a737df621a1c.png)


## Checklist
